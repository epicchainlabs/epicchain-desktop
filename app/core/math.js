// @flow
import BigNumber from 'bignumber.js'

BigNumber.config({ EXPONENTIAL_AT: 12 })

// https://stackoverflow.com/questions/4912788/truncate-not-round-off-decimal-numbers-in-javascript
export const truncateNumber = (num: number, places: number): number =>
  Math.trunc(num * 10 ** places) / 10 ** places

// https://github.com/MikeMcl/bignumber.js/issues/11
export const toBigNumber = (value: number | string) =>
  new BigNumber(String(value))

export const toNumber = (value: string | number) =>
  toBigNumber(value).toNumber()

export const isZero = (amount: string | number) => toBigNumber(amount).equals(0)

export const isNumber = (value: string | number): boolean => {
  try {
    toBigNumber(value)
    return true
  } catch (e) {
    return false
  }
}

export const multiplyNumber = (x: string | number, y: string | number) =>
  toBigNumber(x)
    .mul(toBigNumber(y))
    .toNumber()

export const minusNumber = (x: string | number, y: string | number) =>
  toBigNumber(x)
    .minus(y)
    .toNumber()

export const addNumber = (x: string | number, y: string | number) =>
  toBigNumber(x)
    .plus(y)
    .toNumber()
