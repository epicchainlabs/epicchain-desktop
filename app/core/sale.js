// @flow
export const validateMintTokensInputs = (
  neoToMint: number,
  gasToMint: number,
  scriptHash: string,
  NEO: number,
  GAS: number,
) => {
  let message

  if (neoToMint < 0 || gasToMint < 0 || (neoToMint === 0 && gasToMint === 0)) {
    message = 'You must send positive amounts of NEO or GAS.'
    return [false, message]
  }

  // eslint-disable-next-line radix
  if (neoToMint && parseFloat(neoToMint) !== parseInt(neoToMint)) {
    message = 'You cannot send fractional NEO to a token sale.'
    return [false, message]
  }

  if (
    (neoToMint && Number.isNaN(neoToMint)) ||
    (gasToMint && Number.isNaN(gasToMint))
  ) {
    message = 'Please enter valid numbers only'
    return [false, message]
  }

  if (neoToMint > NEO) {
    message = 'You do not have enough NEO to send.'
    return [false, message]
  }

  if (gasToMint > GAS) {
    message = 'You do not have enough GAS to send.'
    return [false, message]
  }

  if (
    scriptHash.slice(0, 1) !== '0x' &&
    scriptHash.length !== 42 &&
    scriptHash.length !== 40
  ) {
    message = 'Not a valid script hash.'
    return [false, message]
  }

  return [true, '']
}
