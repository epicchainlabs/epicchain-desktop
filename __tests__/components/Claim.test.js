import React from 'react'
import { shallow, mount } from 'enzyme'
import { progressValues } from 'spunky'
import { injectIntl } from 'react-intl'

import { createStore, provideStore } from '../testHelpers'
import Claim from '../../app/containers/Claim/Claim'
import IntlWrapper from '../../app/components/Root/IntlWrapper'
import { DEFAULT_LANGUAGE } from '../../app/core/constants'

const { LOADED } = progressValues

const ClaimWithIntl = injectIntl(Claim)

const initialState = {
  spunky: {
    settings: {
      batch: false,
      progress: LOADED,
      data: {
        language: DEFAULT_LANGUAGE,
      },
      loadedCount: 1,
    },
  },
}

const INTL_STUB = {
  noClaimableGas: 'Address has no claimable GAS',
  claimTimeDisclaimer: 'You can claim GAS once every 5 minutes',
  claimUnavailableInWatch: 'GAS claims are unavailable in Watch mode',
}

describe('Claim', () => {
  const props = {
    doGasClaim: () => {},
    disableClaimButton: false,
    claimAmount: '1.25406935',
  }

  test('should render claim GAS button as enabled', () => {
    const wrapper = shallow(
      <Claim intl={{ formatMessage: id => INTL_STUB[id] }} {...props} />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('should render claim GAS button as disabled', () => {
    const wrapper = shallow(
      <Claim
        intl={{ formatMessage: id => INTL_STUB[id] }}
        {...props}
        disableClaimButton
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('should claim GAS when button is clicked', () => {
    const claimSpy = jest.fn()

    const store = createStore(initialState)
    const wrapper = mount(
      provideStore(
        <IntlWrapper>
          <ClaimWithIntl {...props} doGasClaim={claimSpy} />
        </IntlWrapper>,
        store,
      ),
    )

    wrapper.find('button#claim').simulate('click')

    expect(claimSpy).toHaveBeenCalled()
  })

  test('Claim button should be disabled in watchOnly', () => {
    const claimSpy = jest.fn()
    const watchOnlyProps = { ...props }
    watchOnlyProps.isWatchOnly = true
    const store = createStore(initialState)

    const wrapper = mount(
      provideStore(
        <IntlWrapper>
          <ClaimWithIntl {...watchOnlyProps} doGasClaim={claimSpy} />
        </IntlWrapper>,
        store,
      ),
    )
    expect(claimSpy).toHaveBeenCalledTimes(0)
    expect(wrapper.find('button#claim').prop('disabled')).toBeTruthy()
  })
})
