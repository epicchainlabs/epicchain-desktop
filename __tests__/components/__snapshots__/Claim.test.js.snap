// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claim should render claim GAS button as disabled 1`] = `
<div>
  <Tooltip
    disabled={false}
  >
    <Button
      disabled={true}
      id="claim"
      onClick={[Function]}
      primary={true}
      renderIcon={[Function]}
      shouldCenterButtonLabelText={false}
      type="button"
    >
      <FormattedMessage
        id="dashboardGasClaimButton"
        values={
          Object {
            "amount": "1.25406935",
          }
        }
      />
    </Button>
  </Tooltip>
</div>
`;

exports[`Claim should render claim GAS button as enabled 1`] = `
<div>
  <Tooltip
    disabled={true}
  >
    <Button
      id="claim"
      onClick={[Function]}
      primary={true}
      renderIcon={[Function]}
      shouldCenterButtonLabelText={false}
      type="button"
    >
      <FormattedMessage
        id="dashboardGasClaimButton"
        values={
          Object {
            "amount": "1.25406935",
          }
        }
      />
    </Button>
  </Tooltip>
</div>
`;
